Calendar improvements

chris-y opened the pull request:
Supercedes PR #3 

Improvements include:
 - Gets all events between -4 weeks and +52 weeks when creating the calendar
 - Multi-line details are now formatted as per the specification
 - UID now hashed and unique
 - All day and individual events supported
 - Improved security by requiring calendars to be added to the config with a per-calendar secret
   - BREAKING CHANGE: Calendars not in the config will no longer be accessible (see below)
 - Added ability to assign colours to particular events/calendars
 - Added a friendly name to the calendar (although I haven't found a client which uses it yet)


### Breaking change
Calendars need to be individually specified as accessible, with their own password, to improve security.

```
icalendar:
  secret: yourSuperSecret
```

Needs to be modified to:

```
icalendar:
  calendars:
    - entity_id: calendar.entity_id
      secret: yourSuperSecret
```

Multiple calendars can be specified with their own secrets.  If necessary these can be the same but this should be discouraged.

### Colours

I've added this to the README but in short, events can be assigned colours based on their summary.  This is a case sensitive comparison.

Example from the README which works well with the Waste Collection Schedule integration:

```
icalendar:
  calendars:
    - entity_id: calendar.bins
      secret: yourSuperSecret
      colour: black
  colours:
    - name: "Recycling"
      colour: green
    - name: "Food waste"
      colour: brown
```

I've noticed that Google Calendar appears to ignore the per-event colours but ICSx⁵ brings them across fine.

chris-y committed:
Get all calendar events between -4wks and +52wks
Iterate through them and create a more complete iCalendar file

chris-y committed:
Fix multi-line details

chris-y committed:
Add friendly calendar name

chris-y committed:
Update README

chris-y committed:
Hash the UID as per RFC recommendation

chris-y committed:
Fix duplicate UIDs

chris-y committed:
Add the ability to assign colours to events based on name

chris-y committed:
Document colours

chris-y committed:
Improve security by specifiying valid calendars and passwords manually in config.
Note that each calendar now has a calendar-specific secret, so configuration will need to be migrated from previous versions.
Only named calendars will be available through the /api/ics endpoint.

chris-y committed:
Set a calendar colour

chris-y committed:
Add a note about the line length exceeding the RFC recommendation

chris-y committed:
Ensure times are considered to be UTC

TylonHH commented:
please approve @codyc1515 
Just the 1st upcoming event is useless

chris-y committed:
Use TZ-aware datetime functions to convert to UTC
Some calendars were converting to UTC incorrectly previously

macthaidhg commented:
Just two minor issues to report:

- With the public urls that are generated for the ICS, they import fine into Google Calendar, but they will not import into Outlook (both local app and online versions). To get Outlook working I had to import into Google Calendar, make it publicly available and use the google-generated url for importing into Outlook. Not sure why Outlook prefers Google urls over my self-hosted url?
- After importing into Google Calendar I noticed a +1 hour shift on all events. In Google Calendar, the imported calendar shows “(GMT+00:00) Coordinated Universal Time”. I think the ICS file might need information on Timezone per event. Would it be possible to include this?

chris-y commented:
> * After importing into Google Calendar I noticed a +1 hour shift on all events. In Google Calendar, the imported calendar shows “(GMT+00:00) Coordinated Universal Time”. I think the ICS file might need information on Timezone per event. Would it be possible to include this?

The latest commit fixed a timezone issue, were you using that version?  Prior to that I'd seen the same thing with a local calendar, but it should be converting correctly now.

I'm converting the local time to UTC for the ICS as adding simple offsets is expressly prohibited by the spec - you have to do full timezone selection, and the HA calendar doesn't provide enough information to do that easily.

macthaidhg commented:
> > * After importing into Google Calendar I noticed a +1 hour shift on all events. In Google Calendar, the imported calendar shows “(GMT+00:00) Coordinated Universal Time”. I think the ICS file might need information on Timezone per event. Would it be possible to include this?
> 
> The latest commit fixed a timezone issue, were you using that version? Prior to that I'd seen the same thing with a local calendar, but it should be converting correctly now.
> 
> I'm converting the local time to UTC for the ICS as adding simple offsets is expressly prohibited by the spec - you have to do full timezone selection, and the HA calendar doesn't provide enough information to do that easily.

Thanks, sorry I just see you published a latest commit 2 days ago. Let me trial this and I'll report back!

chris-y committed:
Also add the calendar name as X-WR-CALNAME
I've not found anything which supports the official NAME
Implementations appear to support X-WR-CALNAME instead

chris-y commented:
> * With the public urls that are generated for the ICS, they import fine into Google Calendar, but they will not import into Outlook (both local app and online versions). To get Outlook working I had to import into Google Calendar, make it publicly available and use the google-generated url for importing into Outlook. Not sure why Outlook prefers Google urls over my self-hosted url?

Same here, weird.  There must be something different, but the file looks pretty similar (except start/end times have an extra VALUE= in them), and downloading the file and then adding it as a calendar is working.

Maybe it's an unexpected mimetype or something?

chris-y committed:
Use a common DTSTAMP
On further reading of the spec, DTSTAMP is supposed to be the timestamp
of when the iCalendar object (not the event) was created, which is when
this script runs and is common for all events.
Previously it was using the event start time.

macthaidhg commented:
I trialled your latest commit, but the +1 hour shift still appears for UK/Ireland summer time after importing into Google Calendar. Would you consider adding a configuration option to let users declare a timezone for event export?

Context:

The current UTC-normalized approach is spec-correct, but Google Calendar displays events shifted by +1 hour during daylight saving time. 

Home Assistant’s local calendar doesn’t expose full timezone info per event, so I get why you normalize to UTC - that’s safe. But in real-world use (family calendars, bin collections etc), people would like their events in local time, not UTC.

After checking the generated ICS, I see the following:

- The file has no VTIMEZONE or TZID definitions.
- All events are converted to UTC (Z), even when the instance has a known local timezone.

This causes clients like Google and Outlook to import the feed as UTC and apply an unwanted offset in summer.

Proposal:

Add optional timezone configuration in configuration.yaml, e.g.:

```yaml
icalendar: 
  timezone: "Europe/Dublin" # optional global default 
  calendars: 
    - entity_id: calendar.bins 
      secret: yourSuperSecret 
      colour: black
      timezone: "Europe/Dublin" # optional per-calendar override 
    - entity_id: calendar.family 
      secret: yourSuperSecret 
  colours: 
    - name: "Recycling" 
      colour: green 
    - name: "Food waste" 
      colour: brown 
```

When timezone is defined:

- Output event times as DTSTART;TZID=Europe/Dublin and DTEND;TZID=Europe/Dublin instead of UTC (Z).
- Insert a minimal VTIMEZONE block for the declared zone.
- Otherwise, keep existing UTC behavior so nothing breaks for current users.

This would allow users to opt-in to local timezone handling, eliminating DST-related shifts while maintaining backward compatibility and spec compliance. Would you be open to that as a PR or an approved direction?

chris-y commented:
I created a new local calendar in Home Assistant, and added two events - one before the clocks changed, and one after (I'm in UK so same timezone as you):

<img width="452" height="466" alt="image" src="https://github.com/user-attachments/assets/3fb02b7c-d76b-4694-984d-dfd2e51243f3" />

I looked at the source which has converted the times correctly (the one before the clocks changed is 5PM in UTC)
<img width="370" height="442" alt="image" src="https://github.com/user-attachments/assets/b4656912-b08a-4558-a2ea-acecc4572540" />

I then imported into Google Calendar (I've had to screenshot this and HA calendar several times to ensure these captures are correct as they look so similar!):
<img width="392" height="496" alt="image" src="https://github.com/user-attachments/assets/58171710-47e1-4439-a94f-897f755139f0" />

Times are showing as the correct local times, so at least here everything seems to be working as it should.

Can you run through those steps and show each stage as it might indicate where the problem lies?

chris-y commented:
> Proposal:
> 
> Add optional timezone configuration in configuration.yaml, e.g.:
> 
> ```yaml
> icalendar: 
>   timezone: "Europe/Dublin" # optional global default 
>   calendars: 
>     - entity_id: calendar.bins 
>       secret: yourSuperSecret 
>       colour: black
>       timezone: "Europe/Dublin" # optional per-calendar override 
>     - entity_id: calendar.family 
>       secret: yourSuperSecret 
>   colours: 
>     - name: "Recycling" 
>       colour: green 
>     - name: "Food waste" 
>       colour: brown 
> ```
> 
> When timezone is defined:
> 
>     * Output event times as DTSTART;TZID=Europe/Dublin and DTEND;TZID=Europe/Dublin instead of UTC (Z).
> 
>     * Insert a minimal VTIMEZONE block for the declared zone.
> 
>     * Otherwise, keep existing UTC behavior so nothing breaks for current users.
> 
> 
> This would allow users to opt-in to local timezone handling, eliminating DST-related shifts while maintaining backward compatibility and spec compliance. Would you be open to that as a PR or an approved direction?

In principle, yes.  However note that this isn't my repo or my project.  If we don't hear anything from @codyc1515 I will maintain it on my fork though.

TylonHH commented:
@chris-y how can I use your fork within HA?

chris-y commented:
> @chris-y how can I use your fork within HA?

Hi, given the interest in this and the lack of response from the original dev, I've now merged these changes into main on my fork.  Henceforth, you can now go here: https://github.com/chris-y/ha-icalendar
Click on the "Open HACS repo on HA"
Download and install (ensure it is v1.1 minimum, as that is this updated version)

Any issues, PRs, etc, please raise them on my repo for now.

macthaidhg commented:
> I created a new local calendar in Home Assistant, and added two events - one before the clocks changed, and one after (I'm in UK so same timezone as you):
> 
> <img alt="image" width="452" height="466" src="https://private-user-images.githubusercontent.com/1818803/507101439-3fb02b7c-d76b-4694-984d-dfd2e51243f3.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc3ODE1NjMsIm5iZiI6MTc2Nzc4MTI2MywicGF0aCI6Ii8xODE4ODAzLzUwNzEwMTQzOS0zZmIwMmI3Yy1kNzZiLTQ2OTQtOTg0ZC1kZmQyZTUxMjQzZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDEwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMDdUMTAyMTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTljZjMwZGNlMGI3ZGNkNTA2YjQxZmJmYjdmNmJjYzY3MTA1MTA5YjljZTZhYTU3NjhkMGY1ZGFiZmFmMmY1NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.cAwVW7yf98eEp4Tu5QxNXLZpuHHK5ToBQnhqLIkqqYI">
> I looked at the source which has converted the times correctly (the one before the clocks changed is 5PM in UTC) <img alt="image" width="370" height="442" src="https://private-user-images.githubusercontent.com/1818803/507100363-b4656912-b08a-4558-a2ea-acecc4572540.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc3ODE1NjMsIm5iZiI6MTc2Nzc4MTI2MywicGF0aCI6Ii8xODE4ODAzLzUwNzEwMDM2My1iNDY1NjkxMi1iMDhhLTQ1NTgtYTJlYS1hY2VjYzQ1NzI1NDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDEwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMDdUMTAyMTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGY0M2FjNDVkY2Q2NTY0MDhiYjZkYzk2ZGM1Zjc1MWZjMTI4ZGY5Y2NiNGUxYjJjZGMyZThlOTBkMjgwNzM5YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.sola4wtfmHO5ufYBjOz6_Ure714aqN_BNq5fxzy3Yes">
> 
> I then imported into Google Calendar (I've had to screenshot this and HA calendar several times to ensure these captures are correct as they look so similar!): <img alt="image" width="392" height="496" src="https://private-user-images.githubusercontent.com/1818803/507101848-58171710-47e1-4439-a94f-897f755139f0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc3ODE1NjMsIm5iZiI6MTc2Nzc4MTI2MywicGF0aCI6Ii8xODE4ODAzLzUwNzEwMTg0OC01ODE3MTcxMC00N2UxLTQ0MzktYTk0Zi04OTdmNzU1MTM5ZjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDEwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMDdUMTAyMTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzBlMDRkMTFmNjA2MWRlM2IxZDhhYmZiYWU0MTU1NjUyOTJjZjY0NmQyZGVjYWRmNzYwN2U4YTA0ZjVlZjQ0YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.4yFu252DQEz0DivbkGkrDANGbNJKGZERYh-I5Qnje0g">
> 
> Times are showing as the correct local times, so at least here everything seems to be working as it should.
> 
> Can you run through those steps and show each stage as it might indicate where the problem lies?

Thanks very much for looking into this (apologies for the delayed response, I am only getting around to running the checks now!). I have good news., everything is working perfectly. I ran identical checks to you, entering events in the HA calendar before and after clocks change and when imported into Google calendar and Outlook, they show the correct times. From examining the .ics file, I can confirm exactly what you observed - Home Assistant manages the conversion to UTC, then client apps such as Google Calendar and Outlook convert to GMT. I have no idea why it wasn't working for me previously! Thanks again, and many thanks for your amazing work on this integration. It makes my life so much easier to overlay my HA calendars over my work calendar in Outlook. 

Having trialed this for a few months, I can confirm that this is working perfectly. Just notice a couple of very minor issues with how text is reproduced. Certain characters from the HA calendar don't seem to transfer well to Google Calendar/Outlook. For example, "&" becomes "&amp;" and an apostrophe becomes "&#x27;". Having checked the text file, the erroneous characters appear there, so it is something relating to the generation of the .ics itself.

macthaidhg commented:
> > Proposal:
> > Add optional timezone configuration in configuration.yaml, e.g.:
> > ```yaml
> > icalendar: 
> >   timezone: "Europe/Dublin" # optional global default 
> >   calendars: 
> >     - entity_id: calendar.bins 
> >       secret: yourSuperSecret 
> >       colour: black
> >       timezone: "Europe/Dublin" # optional per-calendar override 
> >     - entity_id: calendar.family 
> >       secret: yourSuperSecret 
> >   colours: 
> >     - name: "Recycling" 
> >       colour: green 
> >     - name: "Food waste" 
> >       colour: brown 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When timezone is defined:
> > ```
> > * Output event times as DTSTART;TZID=Europe/Dublin and DTEND;TZID=Europe/Dublin instead of UTC (Z).
> > 
> > * Insert a minimal VTIMEZONE block for the declared zone.
> > 
> > * Otherwise, keep existing UTC behavior so nothing breaks for current users.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This would allow users to opt-in to local timezone handling, eliminating DST-related shifts while maintaining backward compatibility and spec compliance. Would you be open to that as a PR or an approved direction?
> 
> In principle, yes. However note that this isn't my repo or my project. If we don't hear anything from @codyc1515 I will maintain it on my fork though.

Thanks, now that timezones are converting fine in the client apps, this is no longer necessary!

chris-y commented:
> > I created a new local calendar in Home Assistant, and added two events - one before the clocks changed, and one after (I'm in UK so same timezone as you):
> > <img alt="image" width="452" height="466" src="https://private-user-images.githubusercontent.com/1818803/507101439-3fb02b7c-d76b-4694-984d-dfd2e51243f3.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc3ODE1NjMsIm5iZiI6MTc2Nzc4MTI2MywicGF0aCI6Ii8xODE4ODAzLzUwNzEwMTQzOS0zZmIwMmI3Yy1kNzZiLTQ2OTQtOTg0ZC1kZmQyZTUxMjQzZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDEwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMDdUMTAyMTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTljZjMwZGNlMGI3ZGNkNTA2YjQxZmJmYjdmNmJjYzY3MTA1MTA5YjljZTZhYTU3NjhkMGY1ZGFiZmFmMmY1NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.cAwVW7yf98eEp4Tu5QxNXLZpuHHK5ToBQnhqLIkqqYI">
> > I looked at the source which has converted the times correctly (the one before the clocks changed is 5PM in UTC) <img alt="image" width="370" height="442" src="https://private-user-images.githubusercontent.com/1818803/507100363-b4656912-b08a-4558-a2ea-acecc4572540.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc3ODE1NjMsIm5iZiI6MTc2Nzc4MTI2MywicGF0aCI6Ii8xODE4ODAzLzUwNzEwMDM2My1iNDY1NjkxMi1iMDhhLTQ1NTgtYTJlYS1hY2VjYzQ1NzI1NDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDEwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMDdUMTAyMTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGY0M2FjNDVkY2Q2NTY0MDhiYjZkYzk2ZGM1Zjc1MWZjMTI4ZGY5Y2NiNGUxYjJjZGMyZThlOTBkMjgwNzM5YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.sola4wtfmHO5ufYBjOz6_Ure714aqN_BNq5fxzy3Yes">
> > I then imported into Google Calendar (I've had to screenshot this and HA calendar several times to ensure these captures are correct as they look so similar!): <img alt="image" width="392" height="496" src="https://private-user-images.githubusercontent.com/1818803/507101848-58171710-47e1-4439-a94f-897f755139f0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc3ODE1NjMsIm5iZiI6MTc2Nzc4MTI2MywicGF0aCI6Ii8xODE4ODAzLzUwNzEwMTg0OC01ODE3MTcxMC00N2UxLTQ0MzktYTk0Zi04OTdmNzU1MTM5ZjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDEwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMDdUMTAyMTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzBlMDRkMTFmNjA2MWRlM2IxZDhhYmZiYWU0MTU1NjUyOTJjZjY0NmQyZGVjYWRmNzYwN2U4YTA0ZjVlZjQ0YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.4yFu252DQEz0DivbkGkrDANGbNJKGZERYh-I5Qnje0g">
> > Times are showing as the correct local times, so at least here everything seems to be working as it should.
> > Can you run through those steps and show each stage as it might indicate where the problem lies?
> 
> Thanks very much for looking into this (apologies for the delayed response, I am only getting around to running the checks now!). I have good news., everything is working perfectly. I ran identical checks to you, entering events in the HA calendar before and after clocks change and when imported into Google calendar and Outlook, they show the correct times. From examining the .ics file, I can confirm exactly what you observed - Home Assistant manages the conversion to UTC, then client apps such as Google Calendar and Outlook convert to GMT. I have no idea why it wasn't working for me previously! Thanks again, and many thanks for your amazing work on this integration. It makes my life so much easier to overlay my HA calendars over my work calendar in Outlook.
> 
> Having trialed this for a few months, I can confirm that this is working perfectly. Just notice a couple of very minor issues with how text is reproduced. Certain characters from the HA calendar don't seem to transfer well to Google Calendar/Outlook. For example, "&" becomes "&" and an apostrophe becomes "'". Having checked the text file, the erroneous characters appear there, so it is something relating to the generation of the .ics itself.



> > I created a new local calendar in Home Assistant, and added two events - one before the clocks changed, and one after (I'm in UK so same timezone as you):
> > <img alt="image" width="452" height="466" src="https://private-user-images.githubusercontent.com/1818803/507101439-3fb02b7c-d76b-4694-984d-dfd2e51243f3.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc3ODE1NjMsIm5iZiI6MTc2Nzc4MTI2MywicGF0aCI6Ii8xODE4ODAzLzUwNzEwMTQzOS0zZmIwMmI3Yy1kNzZiLTQ2OTQtOTg0ZC1kZmQyZTUxMjQzZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDEwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMDdUMTAyMTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTljZjMwZGNlMGI3ZGNkNTA2YjQxZmJmYjdmNmJjYzY3MTA1MTA5YjljZTZhYTU3NjhkMGY1ZGFiZmFmMmY1NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.cAwVW7yf98eEp4Tu5QxNXLZpuHHK5ToBQnhqLIkqqYI">
> > I looked at the source which has converted the times correctly (the one before the clocks changed is 5PM in UTC) <img alt="image" width="370" height="442" src="https://private-user-images.githubusercontent.com/1818803/507100363-b4656912-b08a-4558-a2ea-acecc4572540.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc3ODE1NjMsIm5iZiI6MTc2Nzc4MTI2MywicGF0aCI6Ii8xODE4ODAzLzUwNzEwMDM2My1iNDY1NjkxMi1iMDhhLTQ1NTgtYTJlYS1hY2VjYzQ1NzI1NDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDEwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMDdUMTAyMTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGY0M2FjNDVkY2Q2NTY0MDhiYjZkYzk2ZGM1Zjc1MWZjMTI4ZGY5Y2NiNGUxYjJjZGMyZThlOTBkMjgwNzM5YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.sola4wtfmHO5ufYBjOz6_Ure714aqN_BNq5fxzy3Yes">
> > I then imported into Google Calendar (I've had to screenshot this and HA calendar several times to ensure these captures are correct as they look so similar!): <img alt="image" width="392" height="496" src="https://private-user-images.githubusercontent.com/1818803/507101848-58171710-47e1-4439-a94f-897f755139f0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc3ODE1NjMsIm5iZiI6MTc2Nzc4MTI2MywicGF0aCI6Ii8xODE4ODAzLzUwNzEwMTg0OC01ODE3MTcxMC00N2UxLTQ0MzktYTk0Zi04OTdmNzU1MTM5ZjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDEwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMDdUMTAyMTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzBlMDRkMTFmNjA2MWRlM2IxZDhhYmZiYWU0MTU1NjUyOTJjZjY0NmQyZGVjYWRmNzYwN2U4YTA0ZjVlZjQ0YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.4yFu252DQEz0DivbkGkrDANGbNJKGZERYh-I5Qnje0g">
> > Times are showing as the correct local times, so at least here everything seems to be working as it should.
> > Can you run through those steps and show each stage as it might indicate where the problem lies?
> 
> Thanks very much for looking into this (apologies for the delayed response, I am only getting around to running the checks now!). I have good news., everything is working perfectly. I ran identical checks to you, entering events in the HA calendar before and after clocks change and when imported into Google calendar and Outlook, they show the correct times. From examining the .ics file, I can confirm exactly what you observed - Home Assistant manages the conversion to UTC, then client apps such as Google Calendar and Outlook convert to GMT. I have no idea why it wasn't working for me previously! Thanks again, and many thanks for your amazing work on this integration. It makes my life so much easier to overlay my HA calendars over my work calendar in Outlook.
> 
> Having trialed this for a few months, I can confirm that this is working perfectly. Just notice a couple of very minor issues with how text is reproduced. Certain characters from the HA calendar don't seem to transfer well to Google Calendar/Outlook. For example, "&" becomes "&" and an apostrophe becomes "'". Having checked the text file, the erroneous characters appear there, so it is something relating to the generation of the .ics itself.

Cool!  I've raised an issue for that on my repo - for some reason issues were switched off over there.

https://github.com/chris-y/ha-icalendar/issues/1

codyc1515 commented:
Sorry, I never saw the updates on this for some reason. Will close out now as you have forked it.
